token-lending/program/tests/borrow_obligation_liquidity.rs

```diff
@@ -408,3 +411,351 @@ async fn test_borrow_too_large() {
+                vec![usdc_test_reserve.pubkey],
+            ),
+            borrow_obligation_liquidity(
+                spl_token_lending::id(),
+                u64::MAX,
+                Some(SLIPPAGE_LIMIT),
+                sol_test_reserve.liquidity_supply_pubkey,
+                sol_test_reserve.user_liquidity_pubkey,
+                sol_test_reserve.pubkey,
+                sol_test_reserve.liquidity_fee_receiver_pubkey,
+                test_obligation.pubkey,
+                lending_market.pubkey,
+                test_obligation.owner,
+                Some(sol_test_reserve.liquidity_host_pubkey),
+            ),
+        ],
+        Some(&payer.pubkey()),
+    );
+
+    transaction.sign(&[&payer, &user_accounts_owner], recent_blockhash);
+    assert!(banks_client.process_transaction(transaction).await.is_ok());
+
+    let sol_reserve = sol_test_reserve.get_state(&mut banks_client).await;
+    let obligation = test_obligation.get_state(&mut banks_client).await;
+
+    let (total_fee, host_fee) = sol_reserve
+        .config
+        .fees
+        .calculate_borrow_fees(SOL_BORROW_AMOUNT_LAMPORTS.into(), FeeCalculation::Inclusive)
+        .unwrap();
+
+    assert_eq!(total_fee, FEE_AMOUNT);
+    assert_eq!(host_fee, HOST_FEE_AMOUNT);
+
+    let borrow_amount =
+        get_token_balance(&mut banks_client, sol_test_reserve.user_liquidity_pubkey).await;
+    assert_eq!(borrow_amount, SOL_BORROW_AMOUNT_LAMPORTS - FEE_AMOUNT);
+
+    let liquidity = &obligation.borrows[0];
+    assert_eq!(
+        liquidity.borrowed_amount_wads,
+        Decimal::from(SOL_BORROW_AMOUNT_LAMPORTS)
+    );
+
+    let liquidity_supply =
+        get_token_balance(&mut banks_client, sol_test_reserve.liquidity_supply_pubkey).await;
+    assert_eq!(
+        liquidity_supply,
+        initial_liquidity_supply - SOL_BORROW_AMOUNT_LAMPORTS
+    );
+
+    let fee_balance = get_token_balance(
+        &mut banks_client,
+        sol_test_reserve.liquidity_fee_receiver_pubkey,
+    )
+    .await;
+    assert_eq!(fee_balance, FEE_AMOUNT - HOST_FEE_AMOUNT);
+
+    let host_fee_balance =
+        get_token_balance(&mut banks_client, sol_test_reserve.liquidity_host_pubkey).await;
+    assert_eq!(host_fee_balance, HOST_FEE_AMOUNT);
+}
+
+#[tokio::test]
+async fn test_borrow_max_receive_less_than_slippage() {
+    let mut test = ProgramTest::new(
+        "spl_token_lending",
+        spl_token_lending::id(),
+        processor!(process_instruction),
+    );
+
+    // limit to track compute unit increase
+    test.set_compute_max_units(60_000);
+
+    const USDC_DEPOSIT_AMOUNT_FRACTIONAL: u64 =
+        2_000 * FRACTIONAL_TO_USDC * INITIAL_COLLATERAL_RATIO;
+    const SOL_BORROW_AMOUNT_LAMPORTS: u64 = 50 * LAMPORTS_TO_SOL;
+    const USDC_RESERVE_COLLATERAL_FRACTIONAL: u64 = 2 * USDC_DEPOSIT_AMOUNT_FRACTIONAL;
+    const SOL_RESERVE_LIQUIDITY_LAMPORTS: u64 = 2 * SOL_BORROW_AMOUNT_LAMPORTS;
+    const SLIPPAGE_LIMIT: u64 = SOL_BORROW_AMOUNT_LAMPORTS + 1;
```

Comment:
To be at the limit, should this be `SOL_BORROW_AMOUNT_LAMPORTS - FEE_AMOUNT + 1`?